docs/config-validation.md

```diff
@@ -0,0 +1,191 @@
+# Configuration Validation
+
+We use a fork of
+[github.com/go-playground/validator](https://github.com/go-playground/validator)
+to enable validation of our configuration files using struct tags. The validator
+is configured to validate the configuration file at component startup if the
+`-validate` command line flag is set. If the supplied configuration file is
+valid, the component will exit with a zero exit code, otherwise it will exit
+with a non-zero exit code and print the validation errors to stderr.
+
+## Struct Tag Tips
+
+You can find the full list of struct tags supported by the validator [here]
+(https://pkg.go.dev/github.com/go-playground/validator/v10#section-documentation).
+The following are some tips for struct tags that are commonly used in our
+configuration files.
+
+### `required`
+
+This only validates that the value is not the data types default zero value. For
+numbers ensures value is not zero. For strings ensures value is not "". For
+slices, maps, pointers, interfaces, channels and functions ensures the value is
+not nil. Note, this does not validate slices, maps, etc., as they are not nil.
+For these fields you should use `min=1` or `gte=1` to validate the supplied
+values are not empty.
+
+### `omitempty`
+
+The validator will always run any present validations unless the field is also
+tagged with `omitempty`. Said another way, if a validation is present, and that
+validation cannot be satisfied by the zero-value of that field type, then that
+field is also (technically) required. However, when `omitempty` is present,
+validations will only run when the condition of a `required_` (`with`,
+`with_all`, `unless`, etc.) tag is met.
```

Comment:
This confused me -- I thought `omitempty` would just be "verify the other validation tags on this field if it is present, but don't fail if it isn't". Why do "required_with" etc come into it?

Reply:
> The `omitempty` tag allows a field to be empty, or equivalently, to take its zero value. If the field is omitted, none of the other validation tags on the field will be enforced. This can be useful for tags like `validate="omitempty,url"`, for a field which is optional, but must be a URL if it is present.
>
> The `omitempty` tag can be "overruled" by the various conditional required tags. For example, a field with tag `validate="omitempty,url,required_with=Foo"` is allowed to be empty when field Foo is not present, but if field Foo is present, then this field must be present and must be a URL.